File: example/rec_cb.py
"""
Example receiver that uses callbacks for message handling:
- Basic types (Int/Double/Str/Bool) using set_callback()
- Complex type (Example) using set_callback()

This demonstrates the callback approach to receiving messages.
Each message type has its own callback that prints the received value.
"""

import sys
import time

import core
import proto

if __name__ == "__main__":
    table = core.Table(sys.argv, "rec_cb")

    # Set up callbacks for all message types
    table.entry("test_int", proto.Int).set_callback(
        lambda key, msg, timestamp: print(f"int: {msg.val}")
    )
    table.entry("test_double", proto.Double).set_callback(
        lambda key, msg, timestamp: print(f"double: {msg.val}")
    )
    table.entry("test_str", proto.Str).set_callback(
        lambda key, msg, timestamp: print(f"str: {msg.val}")
    )
    table.entry("test_bool", proto.Bool).set_callback(
        lambda key, msg, timestamp: print(f"bool: {msg.val}")
    )
    table.entry("test_msg", proto.Example).set_callback(
        lambda key, msg, timestamp: print(f"example: {msg}")
    )

    try:
        # Keep program running to receive callbacks
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        print("Stopping...")

File: example/cpp/rec_cb.cpp
#include <chrono>
#include <iostream>
#include <thread>
#include <ecal/ecal.h>
#include <ecal/msg/protobuf/subscriber.h>

#include "proto/cpp/basic.pb.h"
#include "proto/cpp/example.pb.h"

int main(int argc, char **argv)
{
    eCAL::Initialize(argc, argv, "rec_cb");

    auto intSub = eCAL::protobuf::CSubscriber<et::proto::Int>("test_int");
    auto doubleSub = eCAL::protobuf::CSubscriber<et::proto::Double>("test_double");
    auto strSub = eCAL::protobuf::CSubscriber<et::proto::Str>("test_str");
    auto boolSub = eCAL::protobuf::CSubscriber<et::proto::Bool>("test_bool");
    auto exampleSub = eCAL::protobuf::CSubscriber<et::proto::Example>("test_msg");

    intSub.AddReceiveCallback([](const char *topic_name_, const et::proto::Int &msg_, long long time_, long long clock_, long long id_)
                              { std::cout << "int: " << msg_.DebugString() << std::endl; });
    doubleSub.AddReceiveCallback([](const char *topic_name_, const et::proto::Double &msg_, long long time_, long long clock_, long long id_)
                                 { std::cout << "double: " << msg_.DebugString() << std::endl; });
    strSub.AddReceiveCallback([](const char *topic_name_, const et::proto::Str &msg_, long long time_, long long clock_, long long id_)
                              { std::cout << "str: " << msg_.DebugString() << std::endl; });
    boolSub.AddReceiveCallback([](const char *topic_name_, const et::proto::Bool &msg_, long long time_, long long clock_, long long id_)
                               { std::cout << "bool: " << msg_.DebugString() << std::endl; });
    exampleSub.AddReceiveCallback([](const char *topic_name_, const et::proto::Example &msg_, long long time_, long long clock_, long long id_)
                                  { std::cout << "example: " << msg_.DebugString() << std::endl; });
    while (eCAL::Ok())
    {
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }
    eCAL::Finalize();
    return 0;
}
